azurelinuxagent/common/protocol/wire.py

```diff
@@ -1033,6 +1044,20 @@ def report_health(self, status, substatus, description):
                                                       resp.read()))
 
     def send_encoded_event(self, provider_id, event_str, encoding='utf8'):
+        """
+        Construct the encoded event and url for telemetry endpoint call
+        Before calling telemetry endpoint, ensure calls are under throttling limits and checks if the number of telemetry endpoint calls 12 in the last 15 seconds
+        (Considered 12 instead actual limit 15 as a conservative approach plus it helps for directly flushing events not hit the throttling issues)
+        (Trade off between delay vs successful event delivery for immediate_flush events)
+        Note: throttling limit is 15 calls in 15 seconds
+        """
+        def can_make_wireserver_call():
```

Comment:
I am not sure we need this logic. I think we just need to keep track the timestamp for the most recent send, then check if WireServer fails with a throttling error and just sleep enough time to go over the 15 secs.

Throttling errors are reported with a 403 or 429 and this response body:

```
<Error xmlns:xsi=\http://www.w3.org/2001/XMLSchema-instance\ xmlns:xsd=\http://www.w3.org/2001/XMLSchema\>
    <Code> RequestRateLimitExceeded </Code>
    <Message> The request was denied because of too many requests by this client. Please retry the request after a few seconds. </Message>\
    <Details> </Details>
</Error>

```

Reply:
well, sleeping 15 secs always when throttling happens adds delay when it's not needed to wait that long. There is a chance that reset could happen before that. For example, we make 15 calls in 14 secs and each call at each min and hit the throttling. My next call would be reset after 1/2 mins instead 15 mins long. This helps for important events as well since we call wireserver directly and agent would wait on it to move forward with next steps. So, I want to deliver as soon as possible.

The new approach looks to me like we want to remediate the problem after we hit issue. but since we know the limits ahead, I went with that approach.

Reply:
Yes, and that is the issue I see. This approach is tightly tied to the current implementation of WireServer and won't work is that changes.

Reply:
The body size is different. It is there to account for the limits in the telemetry infrastructure.

The problem we were trying to solve for throttling is not that we hit the throttling limit, it is that when we hit it, we keep retrying over and over (I think we also increase the number and possibility frequency of the retries) instead of pausing for a little bit and then trying again.

Reply:
You may be able to fix the throttling issue by fixing the logic around

```
DEFAULT_RETRIES = 6
DELAY_IN_SECONDS = 1

THROTTLE_RETRIES = 25
THROTTLE_DELAY_IN_SECONDS = 1

```

Reply:
The other reason we tend to hit the throttling limit is because we accumulate events for 5 minutes and then try to send them all at once, instead of streaming events as they are produced

Reply:
> You may be able to fix the throttling issue by fixing the logic around
> 
> ```
> DEFAULT_RETRIES = 6
> DELAY_IN_SECONDS = 1
> 
> THROTTLE_RETRIES = 25
> THROTTLE_DELAY_IN_SECONDS = 1
> ```

since these limits apply for other endpoints, changing here affect other endpoints as well. I was not so sure why it was mentioned 25 at first place.

Reply:
yes, you need to account for the fact that this code is shared with other requests, but that should not be hard to do

Reply:
> The other reason we tend to hit the throttling limit is because we accumulate events for 5 minutes and then try to send them all at once, instead of streaming events as they are produced

I checked wiresever doc, the other endpoints don't have throttle limit like telemeteyapi. It makes sense why we don't have such logic since we don't know when we hit the throttling but here we know beforehand.  Even we stream I feel, we still need to go within limits because api has ratelimit. 

The other approach looks to me we are not solving the problem, we are just reducing problem